### PR TITLE
[RFC] shell: use msg_outtrans_len_attr for :!cmd

### LINKED ISSUE #7830: :! does not show 2-byte characters
- `nvim --version`: 0.2.2
- Vim (version: ) behaves differently? yes (VIM 7.4)
- Operating system/version: macOS 10.12.6
- Terminal name/version: iTerm2 / 3.1.5
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:!echo "abc 한글 abc"
```

### Actual behaviour

2-byte characters are not shown. 

screen shot 2018-01-09 at 2 03 50 pm

### Expected behaviour

VIM correctly shows all characters

screen shot 2018-01-09 at 2 04 18 pm

### PR DESCRIPTION
**Pull request by bfredl**
This should make multibyte chars work, fixes #7830
Also displays translation for special chars instead of replacing them with whitespace.

### CODE REVIEW

**Comment by bfredl** [comment]
Question: this test is currently failing
```
  -- This matches Vim behavior.
  it('does not capture shell-command output', function()
    eq('\n:!echo "foo"\13\n', funcs.execute('!echo "foo"'))
  end)
```
Is it catastrophic to behave differently from vim here?

**Comment by justinmk** [comment]
> Is it catastrophic to behave differently from vim here?

@bfredl IMO it's ok to break compat here

- it's not likely to be a common case
- there's a workaround
- the existing behavior was an accident of implementation

The bug fixes afforded by this PR are much more important.

**Comment by bfredl** [comment]
Thanks. If it was important it could be hacked in with more save/restore-y `++`/`--`-y things, but if it is not let's leave it.

**Comment by bfredl** [comment]
As far as I can see this also fixes #7788.

**Comment by bfredl** [comment]
More or less done. There is a remaining test issue though: "FileReadPost (using gzip)" in `legacy/011_autocommands_spec.lua` seems to hang, however if I attach screen it doesn't hang, even if I use the default size and even without any `expect()`, which should't really affect any behaviour (unlike screen with non-default size, which will change "press RETURN" issues, or with `expect()` which changes timing), so I don't really know how to debug it...

**Comment by justinmk** [comment]
> "FileReadPost (using gzip)" in legacy/011_autocommands_spec.lua seems to hang

Does nvim_get_mode confirm that a prompt is showing?

**Comment by bfredl** [comment]
Actually the last commit might have solved it. I've have heard that storing non-enum value (even if it is constructed by `|`-ing valid vaules) in an enum typedef is UB, though I would have expected UBSAN to raise error and not randomly trash the value (which seems to happen, so we get into an unexpected branch).

**Comment by bfredl** [comment]
Also given that `os_call_shell` is only invoked once in `call_shell` having both functions seems pretty needless...

**Comment by justinmk** [comment]
`call_shell` does extra stuff that has no place in the os/ layer.

**Comment by bfredl** [comment]
How does my suggestion imply putting the merged function into `os/shell.c`? Rather, `os_call_shell` _also_ seems to  lots of other "extra stuff" in `os/shell.c` that has nothing to do with os, like coping text betwen `curbuf` and invoking `msg_` (which of course an order of magnitude less abstraction breaking than bytebanging the screen buffer as was done before, so this PR is still a net improvement...), the natural consequence would be to _remove_ code from `os/shell.c`, possibly by moving the filter and screen output specific into `ex_cmds.c` or something like that.

**Comment by justinmk** [comment]
Sounds good. Does the throttling still work?

**Comment by bfredl** [comment]
Yes, this doesn't touch the throttling part of the code.

**Comment by bfredl** [comment]
Hmm, but now it works without the enum change, so probably unrelated.

**Comment by bfredl** [comment]
Redirection status (when using filtering, `:w !cmd` etc) was not forwarded for some reason, fixing that now tests pass on travis at least.

Any more comments, or should we merge this? (after all tests pass, of course, qb seems down for the moment)

**Comment by justinmk** [comment]
LGTM!

**Comment by bfredl** [comment]
just waiting for AppVeyor...

**Comment by justinmk** [comment]
This should fix the appveyor failure:

```diff
diff --git a/src/nvim/api/vim.c b/src/nvim/api/vim.c
index 0e7cc428d4ab..30f6e8709d75 100644
--- a/src/nvim/api/vim.c
+++ b/src/nvim/api/vim.c
@@ -239,8 +239,13 @@ String nvim_command_output(String command, Error *err)
   if (capture_local.ga_len > 1) {
     // redir always(?) prepends a newline; remove it.
     char *s = capture_local.ga_data;
+#if defined(WIN32)
+    assert(s[0] == '\r' && s[1] == '\n');
+    memmove(s, s + 2, (size_t)capture_local.ga_len);
+#else
     assert(s[0] == '\n');
     memmove(s, s + 1, (size_t)capture_local.ga_len);
+#endif
     s[capture_local.ga_len - 1] = '\0';
     return (String) {  // Caller will free the memory.
       .data = s,
```

**Comment by bfredl** [comment]
No, that will fail the assert with other message sources that don't prepend the `\r`. Let's just unconditionally turn off translation for `\r`.

**Comment by bfredl** [comment]
This also matches gvim behavior (just checked).

File: test/functional/api/vim_spec.lua [hunk 1]
```diff
@@ -99,8 +100,9 @@ describe('api', function()
           [[echo nvim_command_output('echo "nested1\nnested2"') | ls]]))
     end)
 
-    it('does not return shell |:!| output', function()
-      eq(':!echo "foo"\r\n', nvim('command_output', [[!echo "foo"]]))
+    it('returns shell |:!| output', function()
+      local win_lf = (uname() == 'Windows' and '\r') or ''
+      eq(':!echo foo\r\n\nfoo'..win_lf..'\n', nvim('command_output', [[!echo foo]]))
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua]: any reason not to use `iswin()`? It saves some milliseconds.

Comment (bfredl) [reply test/functional/api/vim_spec.lua]: None other than copypasta some code that also didn't use `iswin()`

**Pull request merged**